cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,858 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+	mesh->origin.x = default_origin.x;
+	mesh->origin.y = default_origin.y;
+	mesh->tiles= (tile_t*)tiles;
+
+	device->mesh = mesh;
+
 	return HB_MC_SUCCESS;	
 }
 
-static int hb_mc_write_tile_reg(uint8_t fd, eva_t eva_id, tile_t *tile, uint32_t epa, uint32_t val) {
-	int error =  hb_mc_copy_to_epa (fd, tile->x, tile->y, epa, &val, 1);
-	if (error != HB_MC_SUCCESS)
-		return HB_MC_FAIL; /* could not memcpy */
-	return HB_MC_SUCCESS;	
+
+
+
+/**
+ * Takes the grid size, tile group dimensions, kernel name, argc, argv* and the finish signal address, calls hb_mc_tile_group_enqueue to initialize all tile groups for grid.
+ * @param[in]  device        Pointer to device
+ * @param[in]  grid_dim      X/Y dimensions of the grid to be initialized
+ * @param[in]  tg_dim        X/Y dimensions of tile groups in grid
+ * @param[in]  name          Kernel name to be executed on tile groups in grid
+ * @param[in]  argc          Number of input arguments to kernel
+ * @param[in]  argv          List of input arguments to kernel
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+int hb_mc_grid_init (device_t *device, hb_mc_dimension_t grid_dim, hb_mc_dimension_t tg_dim, char* name, uint32_t argc, uint32_t argv[]) {
+	int error; 
+	for (hb_mc_idx_t tg_id_x = 0; tg_id_x < grid_dim.x; tg_id_x ++) { 
+		for (hb_mc_idx_t tg_id_y = 0; tg_id_y < grid_dim.y; tg_id_y ++) { 
+			hb_mc_coordinate_t tg_id = hb_mc_coordinate(tg_id_x, tg_id_y);
+			error = hb_mc_tile_group_enqueue(device, device->num_grids, tg_id, grid_dim, tg_dim, name, argc, argv); 
+			if (error != HB_MC_SUCCESS) { 
+				bsg_pr_err("%s: failed to initialize tile group (%d,%d) of grid %d.\n", __func__, tg_id_x, tg_id_y, device->num_grids);
+				return HB_MC_FAIL;
+			}
+		}
+	}
+	device->num_grids ++;
+	return HB_MC_SUCCESS;
 }
 
-/*
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor, which must be obtained from hb_mc_init_host.
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_init (uint8_t *fd, eva_id_t eva_id, char *elf, tile_t *tiles, uint32_t num_tiles) {
-	
-	int error = hb_mc_fifo_init(fd); 
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_init() --> hb_mc_init_host(): failed to initialize host.\n");
+int hb_mc_tile_group_allocate_tiles (device_t *device, tile_group_t *tg){
+	int error;
+	if (tg->dim.x > device->mesh->dim.x){
+		bsg_pr_err("%s: tile group X dimension (%d) larger than mesh X dimension (%d).\n", __func__, tg->dim.x, device->mesh->dim.x);
 		return HB_MC_FAIL;
```

Comment:
You should check `bsg_manycore_errno.h` for more expressive error codes. `HB_MC_INVALID` might be a better choice here.

Reply:
Good catch, fixed!